adapters/repos/db/shard_async_replication.go

```diff
@@ -32,7 +39,279 @@ import (
+					WithField("action", "async_replication").
+					WithField("class_name", s.class.Class).
+					WithField("shard_name", s.name).
+					Info("hashtree initialization stopped")
+				return
+			}
+
+			s.index.logger.
+				WithField("action", "async_replication").
+				WithField("class_name", s.class.Class).
+				WithField("shard_name", s.name).
+				Errorf("iterating objects during hashtree initialization: %v", err)
+			return
+		}
+
+		s.asyncReplicationRWMux.Lock()
+
+		if ctx.Err() != nil {
+			s.index.logger.
+				WithField("action", "async_replication").
+				WithField("class_name", s.class.Class).
+				WithField("shard_name", s.name).
+				Info("hashtree initialization stopped")
+			return
+		}
+
+		s.hashtreeFullyInitialized = true
+
+		s.index.logger.
+			WithField("action", "async_replication").
+			WithField("class_name", s.class.Class).
+			WithField("shard_name", s.name).
+			Info("hashtree successfully initialized")
+
+		s.asyncReplicationRWMux.Unlock()
+
+		s.initHashBeater(ctx)
+	}, s.index.logger)
+
+	return nil
+}
+
+func (s *Shard) mayStopAsyncReplication() {
+	s.asyncReplicationRWMux.Lock()
+	defer s.asyncReplicationRWMux.Unlock()
+
+	if s.hashtree == nil {
+		return
+	}
+
+	s.asyncReplicationCancelFunc()
+
+	if s.hashtreeFullyInitialized {
+		// the hashtree needs to be fully in sync with stored data before it can be persisted
+		s.dumpHashTree()
+	}
+
+	s.hashtree = nil
+	s.hashtreeFullyInitialized = false
+}
+
+func (s *Shard) UpdateAsyncReplication(_ context.Context, enabled bool) error {
+	s.asyncReplicationRWMux.Lock()
+	defer s.asyncReplicationRWMux.Unlock()
+
+	if enabled {
+		if s.hashtree != nil {
+			return nil
+		}
+
+		return s.initAsyncReplication()
+	}
+
+	if s.hashtree == nil {
+		return nil
+	}
+
+	s.asyncReplicationCancelFunc()
+
+	s.hashtree = nil
+	s.hashtreeFullyInitialized = false
+
+	return nil
+}
+
+func (s *Shard) buildCompactHashTree() (hashtree.AggregatedHashTree, error) {
+	return hashtree.NewCompactHashTree(math.MaxUint64, 16)
+}
+
+func (s *Shard) dumpHashTree() error {
+	var b [8]byte
+	binary.BigEndian.PutUint64(b[:], uint64(time.Now().UnixNano()))
+
+	hashtreeFilename := filepath.Join(s.pathHashTree(), fmt.Sprintf("hashtree-%x.ht", string(b[:])))
+
+	f, err := os.OpenFile(hashtreeFilename, os.O_CREATE|os.O_WRONLY|os.O_APPEND, os.ModePerm)
+	if err != nil {
+		return fmt.Errorf("storing hashtree in %q: %w", hashtreeFilename, err)
+	}
+	defer f.Close()
+
+	w := bufio.NewWriter(f)
+
+	_, err = s.hashtree.Serialize(w)
+	if err != nil {
+		return fmt.Errorf("storing hashtree in %q: %w", hashtreeFilename, err)
+	}
+
+	err = w.Flush()
+	if err != nil {
+		return fmt.Errorf("storing hashtree in %q: %w", hashtreeFilename, err)
+	}
+
+	err = f.Sync()
+	if err != nil {
+		return fmt.Errorf("storing hashtree in %q: %w", hashtreeFilename, err)
+	}
+
+	return nil
+}
+
+func (s *Shard) HashTreeLevel(ctx context.Context, level int, discriminant *hashtree.Bitset) (digests []hashtree.Digest, err error) {
```

Comment:
this method needs to be public?

Reply:
yes, this method is exposed, weaviate nodes holding the same shard will exchange that info, and this method is ultimately exposed as an endpoint